internal/util/ssa/request_tracker.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssa
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"sigs.k8s.io/cluster-api/internal/util/hash"
+)
+
+const (
+	// ttl is the duration for which we keep the objects in the tracker.
+	ttl = 10 * time.Minute
+
+	// expirationInterval is the interval in which we will remove expired objects
+	// from the tracker.
+	expirationInterval = 10 * time.Hour
+)
+
+// RequestTracker tracks request results.
+type RequestTracker interface {
```

Comment:
What about calling this (ssa.)CacheProxy?
At the end goal of this component is to prevent calls to the backed by storing some data locally, like a [web proxy cache](https://www.techopedia.com/definition/2453/web-proxy-cache) does

Reply:
I think the Proxy part probably only makes sense if we run the requests through this object as well? Right now I kept it as a separate component